File: ikv-go-client/reader.go
package ikvclient

import (
	"errors"
	"fmt"

	schemas "github.com/inlinedio/ikv-store/ikv-go-client/schemas"
	"google.golang.org/protobuf/proto"
)

type DefaultIKVReader struct {
	clientoptions *ClientOptions
	native_reader *NativeReaderV2
}

func NewDefaultIKVReader(clientOptions *ClientOptions) (IKVReader, error) {
	if clientOptions == nil {
		return nil, errors.New("clientOptions are required")
	}

	// no assertion on required options
	// will be done by native call
	return &DefaultIKVReader{
		clientoptions: clientOptions,
		native_reader: nil,
	}, nil
}

// Startup. Reader fetches and combines server/client configs
// and opens embedded index via cgo.
func (reader *DefaultIKVReader) Startup() error {
	// dynamic load native IKV binaries
	if reader.native_reader == nil {
		mountdir, exists := reader.clientoptions.config.StringConfigs["mount_directory"]
		if !exists {
			return errors.New("mount_directory is a required client specified option")
		}

		bin_manager, err := NewBinaryManager(mountdir)
		if err != nil {
			return err
		}

		dll_path, err := bin_manager.GetPathToNativeBinary()
		if err != nil {
			return err
		}

		reader.native_reader, err = NewNativeReaderV2(dll_path)
		if err != nil {
			return err
		}
	}

	// fetch server supplied options, and override them with client options
	config, err := reader.createIKVConfig()
	if err != nil {
		return err
	}
	config_bytes, err := proto.Marshal(config)
	if err != nil {
		return err
	}

	// open embedded index reader
	err = reader.native_reader.Open(config_bytes)
	if err != nil {
		return fmt.Errorf("cannot initialize reader: %w", err)
	}

	return nil
}

// Shutdown. Reader invokes shutdown sequence on the embedded index
// via cgo.
func (reader *DefaultIKVReader) Shutdown() error {
	if err := reader.native_reader.Close(); err != nil {
		return err
	}

	return nil
}

func (reader *DefaultIKVReader) HealthCheck() (bool, error) {
	return reader.native_reader.HealthCheck("healthcheck")
}

func (reader *DefaultIKVReader) GetBytesValue(key interface{}, fieldname string) ([]byte, error) {
	switch primaryKey := key.(type) {
	case string:
		return reader.native_reader.GetFieldValue(
			[]byte(primaryKey),
			fieldname), nil
	case []byte:
		return reader.native_reader.GetFieldValue(
			primaryKey,
			fieldname), nil
	default:
		return nil, errors.New("key can only be a string or []byte")
	}
}

func (reader *DefaultIKVReader) GetStringValue(key interface{}, fieldname string) (string, error) {
	bv, err := reader.GetBytesValue(key, fieldname)
	if err != nil {
		return "", err
	}

	return string(bv), nil
}

func (reader *DefaultIKVReader) createIKVConfig() (*schemas.IKVStoreConfig, error) {
	client, err := NewDefaultIKVWriter(reader.clientoptions)
	if err != nil {
		return nil, fmt.Errorf("cannot fetch server supplied options: %w", err)
	}

	err = client.Startup()
	if err != nil {
		return nil, fmt.Errorf("cannot fetch server supplied options: %w", err)
	}

	config, err := client.serverSuppliedConfig()
	if err != nil {
		return nil, fmt.Errorf("cannot fetch server supplied options: %w", err)
	}

	err = client.Shutdown()
	if err != nil {
		return nil, fmt.Errorf("cannot fetch server supplied options: %w", err)
	}

	if config.StringConfigs == nil {
		config.StringConfigs = make(map[string]string)
	}
	for k, v := range reader.clientoptions.config.StringConfigs {
		config.StringConfigs[k] = v
	}

	if config.IntConfigs == nil {
		config.IntConfigs = make(map[string]int64)
	}
	for k, v := range reader.clientoptions.config.IntConfigs {
		config.IntConfigs[k] = v
	}

	if config.FloatConfigs == nil {
		config.FloatConfigs = make(map[string]float32)
	}
	for k, v := range reader.clientoptions.config.FloatConfigs {
		config.FloatConfigs[k] = v
	}

	if config.BytesConfigs == nil {
		config.BytesConfigs = make(map[string][]byte)
	}
	for k, v := range reader.clientoptions.config.BytesConfigs {
		config.BytesConfigs[k] = v
	}

	if config.BooleanConfigs == nil {
		config.BooleanConfigs = make(map[string]bool)
	}
	for k, v := range reader.clientoptions.config.BooleanConfigs {
		config.BooleanConfigs[k] = v
	}

	return config, nil
}

File: ikv-python-client/src/ikvpy/reader.py
from typing import List, Optional, Iterator
from client import IKVReader
from clientoptions import ClientOptions
from writer import IKVWriterImpl
from bin_manager import NativeBinaryManager
from utils import is_valid_str_or_raise

import native_reader
import schemas.common_pb2 as common_pb2

class IKVReaderImpl(IKVReader):
    def __init__(self, client_options: ClientOptions):
        """ See client.py for usage documentation. """
        if client_options is None:
            raise TypeError("client_options are required and can't be None")
        self.client_options = client_options
        self.native_reader = None

    def startup(self):
        """ See client.py for usage documentation. """
        # download dll and initialize native reader
        mount_dir = is_valid_str_or_raise(self.client_options.get_ikv_config().stringConfigs["mount_directory"])
        bin_manager = NativeBinaryManager(mount_dir=mount_dir)
        dll_path = bin_manager.get_path_to_dll()
        if dll_path is None: raise RuntimeError("Cannot download IKV native binary")
        self.native_reader = native_reader.NativeReader(dll_path)

        # fetch server supplied config and merge with client cfg
        writer: IKVWriterImpl = IKVWriterImpl(self.client_options)
        writer.startup()
        server_cfg = writer.fetch_server_supplied_config()
        writer.shutdown()
        if server_cfg is None: raise RuntimeError("cannot fetch startup cfg from IKV cloud for given client options")
        merged_cfg = self._merge_configs(server_cfg)
        ikv_config_bytes = bytes(merged_cfg.SerializeToString())
        
        # open embedded db
        self.native_reader.open(ikv_config_bytes)

    def shutdown(self):
        self.native_reader.close()

    def get_bytes_value(self, primary_key, field_name: str) -> Optional[bytes]:
        """ See client.py for usage documentation. """
        if isinstance(primary_key, str):
            return self.native_reader.get_bytes_field_value(bytes(primary_key.encode('utf-8')), field_name)
        
        if isinstance(primary_key, bytes):
            return self.native_reader.get_bytes_field_value(primary_key, field_name)
        
        # ffi.new() does not accept bytearray type as is
        if isinstance(primary_key, bytearray):
            return self.native_reader.get_bytes_field_value(bytes(primary_key), field_name)
        
        raise TypeError("unsupported primary_key type: {}, supported: str/bytes/bytearray".format(type(primary_key)))

    def get_string_value(self, primary_key, field_name: str) -> Optional[str]:
        """ See client.py for usage documentation. """
        if isinstance(primary_key, str):
            return self.native_reader.get_string_field_value(bytes(primary_key.encode('utf-8')), field_name)
        
        if isinstance(primary_key, bytes):
            return self.native_reader.get_string_field_value(primary_key, field_name)
        
        # ffi.new() does not accept bytearray type as is
        if isinstance(primary_key, bytearray):
            return self.native_reader.get_string_field_value(bytes(primary_key), field_name)

        raise TypeError("unsupported primary_key type: {}, supported: str/bytes/bytearray".format(type(primary_key)))

    def multiget_bytes_values(self, bytes_primary_keys: List[bytes] = [], str_primary_keys: List[str] = [],\
            field_names: List[str] = []) -> Iterator[Optional[bytes]]:
        return self.native_reader.multiget_bytes_field_values(bytes_primary_keys, str_primary_keys, field_names)

    def multiget_string_values(self, bytes_primary_keys: List[bytes] = [], str_primary_keys: List[str] = [],\
            field_names: List[str] = []) -> Iterator[Optional[str]]:
        return self.native_reader.multiget_str_field_values(bytes_primary_keys, str_primary_keys, field_names)

    def _merge_configs(self, server_cfg: common_pb2.IKVStoreConfig) -> common_pb2.IKVStoreConfig:
        """ Internal method, overrides & merges server supplied cfg with client supplier cfg. """
        client_cfg = self.client_options.get_ikv_config()

        string_configs = server_cfg.stringConfigs if server_cfg.stringConfigs is not None else {}
        if client_cfg.stringConfigs is not None:
            for k,v in client_cfg.stringConfigs.items():
                string_configs[k] = v

        intConfigs = server_cfg.intConfigs if server_cfg.intConfigs is not None else {}
        if client_cfg.intConfigs is not None:
            for k,v in client_cfg.intConfigs.items():
                intConfigs[k] = v

        floatConfigs = server_cfg.floatConfigs if server_cfg.floatConfigs is not None else {}
        if client_cfg.floatConfigs is not None:
            for k,v in client_cfg.floatConfigs.items():
                floatConfigs[k] = v

        bytesConfigs = server_cfg.bytesConfigs if server_cfg.bytesConfigs is not None else {}
        if client_cfg.bytesConfigs is not None:
            for k,v in client_cfg.bytesConfigs.items():
                bytesConfigs[k] = v

        booleanConfigs = server_cfg.booleanConfigs if server_cfg.booleanConfigs is not None else {}
        if client_cfg.booleanConfigs is not None:
            for k,v in client_cfg.booleanConfigs.items():
                booleanConfigs[k] = v

        return common_pb2.IKVStoreConfig(stringConfigs=string_configs, intConfigs=intConfigs,\
                floatConfigs=floatConfigs, bytesConfigs=bytesConfigs, booleanConfigs=booleanConfigs)